File: apps/mobile/lib/markets/estimate.dart
import 'validation.dart';

const researchAaplxMint = 'XsbEhLAtcf6HdfpFZ5xEMdqW8nfAvcsP5bdudRLJzJp';
const researchUsdcMint = 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v';

enum StockEstimateSide { buy, sell }

class StockEstimateRequest {
  const StockEstimateRequest({
    required this.assetId,
    required this.variantMint,
    required this.side,
    required this.amountRaw,
  });
  final String assetId, variantMint, amountRaw;
  final StockEstimateSide side;
  void validate() {
    try {
      if (assetId != 'apple' || variantMint != researchAaplxMint) {
        researchInvalid();
      }
      researchRawAmount(amountRaw);
      if (BigInt.parse(amountRaw) > BigInt.from(100000000)) researchInvalid();
    } catch (_) {
      throw const StockResearchException('MARKET_INPUT_INVALID');
    }
  }

  Map<String, String> get queryParameters => {
    'assetId': assetId,
    'variantMint': variantMint,
    'side': side.name,
    'amountRaw': amountRaw,
  };
  @override
  bool operator ==(Object other) =>
      other is StockEstimateRequest &&
      assetId == other.assetId &&
      variantMint == other.variantMint &&
      side == other.side &&
      amountRaw == other.amountRaw;
  @override
  int get hashCode => Object.hash(assetId, variantMint, side, amountRaw);
}

class StockEstimateInput {
  StockEstimateInput._(this.symbol, this.mint, this.decimals, this.amountRaw);
  factory StockEstimateInput.fromJson(Object? value) {
    final data = researchObject(value, const {
      'symbol',
      'mint',
      'decimals',
      'amountRaw',
    });
    return StockEstimateInput._(
      researchText(data['symbol'], 40),
      researchMint(data['mint']),
      researchInteger(data['decimals'], 0, 255),
      researchRawAmount(data['amountRaw']),
    );
  }
  final String symbol, mint, amountRaw;
  final int decimals;
}

class StockEstimateOutput {
  StockEstimateOutput._(
    this.symbol,
    this.mint,
    this.decimals,
    this.estimatedAmountRaw,
    this.quotedMinimumAmountRaw,
  );
  factory StockEstimateOutput.fromJson(Object? value) {
    final data = researchObject(value, const {
      'symbol',
      'mint',
      'decimals',
      'estimatedAmountRaw',
      'quotedMinimumAmountRaw',
    });
    return StockEstimateOutput._(
      researchText(data['symbol'], 40),
      researchMint(data['mint']),
      researchInteger(data['decimals'], 0, 255),
      researchRawAmount(data['estimatedAmountRaw']),
      researchRawAmount(data['quotedMinimumAmountRaw']),
    );
  }
  final String symbol, mint, estimatedAmountRaw, quotedMinimumAmountRaw;
  final int decimals;
}

class StockEstimate {
  StockEstimate._(
    this.request,
    this.input,
    this.output,
    this.slippageBps,
    this.swapFeeBasisPoints,
    this.swapFeeMint,
    this.router,
    this.requestedAt,
    this.receivedAt,
    this.refreshAfter,
    this.providerExpiresAt,
  );
  factory StockEstimate.fromJson(Object? value) {
    final data = researchObject(value, const {
      'schemaVersion',
      'kind',
      'network',
      'provider',
      'executable',
      'walletChecked',
      'networkFees',
      'input',
      'output',
      'slippageBps',
      'swapFee',
      'router',
      'requestedAt',
      'receivedAt',
      'refreshAfter',
      'providerExpiresAt',
      'assetId',
      'variantMint',
      'side',
      'executionEnabled',
      'eligibility',
      'amountUnits',
    });
    researchSchema(data['schemaVersion']);
    if (data['kind'] != 'indicative' ||
        data['network'] != 'solana:mainnet-beta' ||
        data['provider'] != 'jupiter-swap-v2' ||
        data['executable'] != false ||
        data['walletChecked'] != false ||
        data['networkFees'] != null ||
        data['executionEnabled'] != false ||
        data['eligibility'] != 'unverified' ||
        data['amountUnits'] != 'raw_token_units') {
      researchInvalid();
    }
    final side = StockEstimateSide.values
        .where((side) => side.name == data['side'])
        .firstOrNull;
    if (side == null) researchInvalid();
    final input = StockEstimateInput.fromJson(data['input']);
    final output = StockEstimateOutput.fromJson(data['output']);
    final request = StockEstimateRequest(
      assetId: researchAssetId(data['assetId']),
      variantMint: researchMint(data['variantMint']),
      side: side,
      amountRaw: input.amountRaw,
    );
    if (request.assetId != 'apple' ||
        request.variantMint != researchAaplxMint ||
        BigInt.parse(input.amountRaw) > BigInt.from(100000000)) {
      researchInvalid();
    }
    final buying = side == StockEstimateSide.buy;
    if (input.mint != (buying ? researchUsdcMint : researchAaplxMint) ||
        input.decimals != (buying ? 6 : 8) ||
        input.symbol != (buying ? 'USDC' : 'AAPLx') ||
        output.mint != (buying ? researchAaplxMint : researchUsdcMint) ||
        output.decimals != (buying ? 8 : 6) ||
        output.symbol != (buying ? 'AAPLx' : 'USDC')) {
      researchInvalid();
    }
    final slippage = researchInteger(data['slippageBps'], 0, 10000);
    final amount = BigInt.parse(output.estimatedAmountRaw);
    final minimum = BigInt.parse(output.quotedMinimumAmountRaw);
    if (minimum > amount ||
        minimum <
            amount * BigInt.from(10000 - slippage) ~/ BigInt.from(10000)) {
      researchInvalid();
    }
    final fee = researchObject(data['swapFee'], const {'basisPoints', 'mint'});
    final feeMint = researchMint(fee['mint']);
    if (feeMint != input.mint && feeMint != output.mint) researchInvalid();
    final router = researchText(data['router'], 40);
    if (!const {'metis', 'jupiterz', 'dflow', 'okx'}.contains(router)) {
      researchInvalid();
    }
    final requested = researchTimestamp(data['requestedAt']);
    final received = researchTimestamp(data['receivedAt']);
    final refresh = researchTimestamp(data['refreshAfter']);
    final expires = data['providerExpiresAt'] == null
        ? null
        : researchTimestamp(data['providerExpiresAt']);
    if (DateTime.parse(received).isBefore(DateTime.parse(requested)) ||
        !DateTime.parse(refresh).isAfter(DateTime.parse(received)) ||
        DateTime.parse(refresh).difference(DateTime.parse(requested)) >
            const Duration(seconds: 10) ||
        (expires != null &&
            DateTime.parse(refresh).isAfter(DateTime.parse(expires)))) {
      researchInvalid();
    }
    return StockEstimate._(
      request,
      input,
      output,
      slippage,
      researchInteger(fee['basisPoints'], 0, 10000),
      feeMint,
      router,
      requested,
      received,
      refresh,
      expires,
    );
  }
  final StockEstimateRequest request;
  final StockEstimateInput input;
  final StockEstimateOutput output;
  final int slippageBps, swapFeeBasisPoints;
  final String swapFeeMint, router, requestedAt, receivedAt, refreshAfter;
  final String? providerExpiresAt;
  String get kind => 'indicative';
  String get provider => 'jupiter-swap-v2';
  String get network => 'solana:mainnet-beta';
  String get amountUnits => 'raw_token_units';
  String get eligibility => 'unverified';
  bool get executable => false;
  bool get executionEnabled => false;
  bool get walletChecked => false;
  Object? get networkFees => null;
  bool needsRefresh(DateTime now) =>
      !now.isBefore(DateTime.parse(refreshAfter));
}

File: apps/web/src/markets/estimate.ts
import {assetId, integer, invalid, mint, rawAmount, record, schema, text, timestamp, StockResearchError} from './validation.js';

export const RESEARCH_AAPLX_MINT = 'XsbEhLAtcf6HdfpFZ5xEMdqW8nfAvcsP5bdudRLJzJp';
export const RESEARCH_USDC_MINT = 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v';
export interface StockEstimateRequest {
  readonly assetId: 'apple'; readonly variantMint: typeof RESEARCH_AAPLX_MINT;
  readonly side: 'buy' | 'sell'; readonly amountRaw: string;
}
export interface StockEstimate {
  readonly schemaVersion: 1; readonly kind: 'indicative'; readonly network: 'solana:mainnet-beta';
  readonly provider: 'jupiter-swap-v2'; readonly executable: false; readonly walletChecked: false; readonly networkFees: null;
  readonly input: Readonly<{symbol: 'USDC' | 'AAPLx'; mint: string; decimals: 6 | 8; amountRaw: string}>;
  readonly output: Readonly<{symbol: 'USDC' | 'AAPLx'; mint: string; decimals: 6 | 8; estimatedAmountRaw: string; quotedMinimumAmountRaw: string}>;
  readonly slippageBps: number; readonly swapFee: Readonly<{basisPoints: number; mint: string}>;
  readonly router: 'metis' | 'jupiterz' | 'dflow' | 'okx';
  readonly requestedAt: string; readonly receivedAt: string; readonly refreshAfter: string; readonly providerExpiresAt: string | null;
  readonly assetId: 'apple'; readonly variantMint: typeof RESEARCH_AAPLX_MINT; readonly side: 'buy' | 'sell';
  readonly executionEnabled: false; readonly eligibility: 'unverified'; readonly amountUnits: 'raw_token_units';
}
export function parseStockEstimateRequest(value: unknown): StockEstimateRequest {
  try {
    const data = record(value, ['assetId', 'variantMint', 'side', 'amountRaw']);
    if (data['assetId'] !== 'apple' || data['variantMint'] !== RESEARCH_AAPLX_MINT || (data['side'] !== 'buy' && data['side'] !== 'sell')) invalid();
    const amountRaw = rawAmount(data['amountRaw']); if (BigInt(amountRaw) > 100000000n) invalid();
    return Object.freeze({assetId: 'apple', variantMint: RESEARCH_AAPLX_MINT, side: data['side'], amountRaw});
  } catch { throw new StockResearchError('MARKET_INPUT_INVALID'); }
}
export function parseStockEstimate(value: unknown): StockEstimate {
  const data = record(value, ['schemaVersion', 'kind', 'network', 'provider', 'executable', 'walletChecked', 'networkFees',
    'input', 'output', 'slippageBps', 'swapFee', 'router', 'requestedAt', 'receivedAt', 'refreshAfter', 'providerExpiresAt',
    'assetId', 'variantMint', 'side', 'executionEnabled', 'eligibility', 'amountUnits']);
  schema(data['schemaVersion']);
  if (data['kind'] !== 'indicative' || data['network'] !== 'solana:mainnet-beta' || data['provider'] !== 'jupiter-swap-v2' ||
      data['executable'] !== false || data['walletChecked'] !== false || data['networkFees'] !== null ||
      data['executionEnabled'] !== false || data['eligibility'] !== 'unverified' || data['amountUnits'] !== 'raw_token_units') invalid();
  const input = record(data['input'], ['symbol', 'mint', 'decimals', 'amountRaw']);
  const output = record(data['output'], ['symbol', 'mint', 'decimals', 'estimatedAmountRaw', 'quotedMinimumAmountRaw']);
  const side = data['side']; if (side !== 'buy' && side !== 'sell') invalid();
  if (assetId(data['assetId']) !== 'apple' || data['variantMint'] !== RESEARCH_AAPLX_MINT) invalid();
  const amountRaw = rawAmount(input['amountRaw']); if (BigInt(amountRaw) > 100000000n) invalid();
  const buying = side === 'buy';
  const inputMint = buying ? RESEARCH_USDC_MINT : RESEARCH_AAPLX_MINT;
  const outputMint = buying ? RESEARCH_AAPLX_MINT : RESEARCH_USDC_MINT;
  const inputSymbol = buying ? 'USDC' : 'AAPLx'; const outputSymbol = buying ? 'AAPLx' : 'USDC';
  const inputDecimals = buying ? 6 : 8; const outputDecimals = buying ? 8 : 6;
  if (input['symbol'] !== inputSymbol || input['mint'] !== inputMint || input['decimals'] !== inputDecimals ||
      output['symbol'] !== outputSymbol || output['mint'] !== outputMint || output['decimals'] !== outputDecimals) invalid();
  const estimatedAmountRaw = rawAmount(output['estimatedAmountRaw']), quotedMinimumAmountRaw = rawAmount(output['quotedMinimumAmountRaw']);
  const slippageBps = integer(data['slippageBps'], 0, 10000);
  if (BigInt(quotedMinimumAmountRaw) > BigInt(estimatedAmountRaw) ||
      BigInt(quotedMinimumAmountRaw) < BigInt(estimatedAmountRaw) * BigInt(10000 - slippageBps) / 10000n) invalid();
  const fee = record(data['swapFee'], ['basisPoints', 'mint']); const feeMint = mint(fee['mint']);
  if (feeMint !== inputMint && feeMint !== outputMint) invalid();
  const router = text(data['router'], 40); if (!['metis', 'jupiterz', 'dflow', 'okx'].includes(router)) invalid();
  const requestedAt = timestamp(data['requestedAt']), receivedAt = timestamp(data['receivedAt']), refreshAfter = timestamp(data['refreshAfter']);
  const providerExpiresAt = data['providerExpiresAt'] === null ? null : timestamp(data['providerExpiresAt']);
  if (Date.parse(receivedAt) < Date.parse(requestedAt) || Date.parse(refreshAfter) <= Date.parse(receivedAt) ||
      Date.parse(refreshAfter) - Date.parse(requestedAt) > 10000 ||
      providerExpiresAt !== null && Date.parse(refreshAfter) > Date.parse(providerExpiresAt)) invalid();
  return Object.freeze({schemaVersion: 1, kind: 'indicative', network: 'solana:mainnet-beta', provider: 'jupiter-swap-v2',
    executable: false, walletChecked: false, networkFees: null,
    input: Object.freeze({symbol: inputSymbol, mint: inputMint, decimals: inputDecimals, amountRaw}),
    output: Object.freeze({symbol: outputSymbol, mint: outputMint, decimals: outputDecimals, estimatedAmountRaw, quotedMinimumAmountRaw}),
    slippageBps, swapFee: Object.freeze({basisPoints: integer(fee['basisPoints'], 0, 10000), mint: feeMint}), router: router as StockEstimate['router'],
    requestedAt, receivedAt, refreshAfter, providerExpiresAt, assetId: 'apple', variantMint: RESEARCH_AAPLX_MINT,
    side, executionEnabled: false, eligibility: 'unverified', amountUnits: 'raw_token_units'});
}
